File: examples/plugins/example/reverse.py
#!/usr/bin/env python3
"""
Reverse String Plugin - Example Executable Plugin

This demonstrates how to create an executable plugin in any language.
The plugin receives JSON via stdin and outputs JSON via stdout.
"""

import json
import sys


def reverse_string(text: str, preserve_case: bool = True) -> dict:
    """
    Reverse a string, optionally preserving case.

    Args:
        text: The string to reverse
        preserve_case: Whether to preserve character casing

    Returns:
        Dictionary with success status and result
    """
    try:
        if preserve_case:
            # Simple reversal
            reversed_text = text[::-1]
        else:
            # Reverse and convert to lowercase
            reversed_text = text[::-1].lower()

        return {
            'success': True,
            'original': text,
            'reversed': reversed_text,
            'length': len(text),
            'preserved_case': preserve_case
        }
    except Exception as e:
        return {
            'success': False,
            'error': f'Failed to reverse string: {str(e)}',
            'error_type': 'processing_error'
        }


def main():
    """Main entry point - reads JSON from stdin, outputs JSON to stdout."""
    try:
        # Read input from stdin
        input_data = json.loads(sys.stdin.read())

        # Extract parameters
        text = input_data.get('text', '')
        preserve_case = input_data.get('preserve_case', True)

        # Validate input
        if not text:
            result = {
                'success': False,
                'error': 'Missing required parameter: text',
                'error_type': 'validation_error'
            }
        else:
            # Execute the operation
            result = reverse_string(text, preserve_case)

        # Output result as JSON
        print(json.dumps(result))

    except json.JSONDecodeError as e:
        # Invalid JSON input
        error_result = {
            'success': False,
            'error': f'Invalid JSON input: {str(e)}',
            'error_type': 'input_error'
        }
        print(json.dumps(error_result))
        sys.exit(1)

    except Exception as e:
        # Unexpected error
        error_result = {
            'success': False,
            'error': f'Unexpected error: {str(e)}',
            'error_type': 'system_error'
        }
        print(json.dumps(error_result))
        sys.exit(1)


if __name__ == '__main__':
    main()

File: examples/plugins/example-node/reverse.ts
#!/usr/bin/env node
/**
 * Reverse String Plugin - Example Executable Plugin (TypeScript/Node.js)
 *
 * This demonstrates how to create an executable plugin in TypeScript/Node.js.
 * The plugin receives JSON via stdin and outputs JSON via stdout.
 * Configuration is read from environment variables.
 *
 * Signal Handling:
 * - Handles SIGTERM/SIGINT for graceful shutdown when interrupted
 * - More complex plugins should cleanup resources (files, connections, etc.) in the handler
 */

interface InputData {
  text: string;
  preserve_case?: boolean;
}

interface SuccessResult {
  success: true;
  original: string;
  reversed: string;
  length: number;
  preserved_case: boolean;
  config_applied: {
    prefix: string | null;
    suffix: string | null;
    api_key_present: boolean;
  };
}

interface ErrorResult {
  success: false;
  error: string;
  error_type: string;
}

type Result = SuccessResult | ErrorResult;

/**
 * Reverse a string, optionally preserving case.
 * Applies prefix/suffix from configuration if provided.
 */
function reverseString(text: string, preserveCase: boolean = true): Result {
  try {
    // Read configuration from environment variables
    const prefix = process.env.PLUGIN_CONFIG_PREFIX || '';
    const suffix = process.env.PLUGIN_CONFIG_SUFFIX || '';
    const apiKey = process.env.PLUGIN_CONFIG_API_KEY || '';

    let reversedText: string;

    if (preserveCase) {
      // Simple reversal
      reversedText = text.split('').reverse().join('');
    } else {
      // Reverse and convert to lowercase
      reversedText = text.split('').reverse().join('').toLowerCase();
    }

    // Apply prefix and suffix from config
    if (prefix) {
      reversedText = prefix + reversedText;
    }
    if (suffix) {
      reversedText = reversedText + suffix;
    }

    return {
      success: true,
      original: text,
      reversed: reversedText,
      length: text.length,
      preserved_case: preserveCase,
      config_applied: {
        prefix: prefix || null,
        suffix: suffix || null,
        api_key_present: Boolean(apiKey)
      }
    };
  } catch (e) {
    const error = e as Error;
    return {
      success: false,
      error: `Failed to reverse string: ${error.message}`,
      error_type: 'processing_error'
    };
  }
}

/**
 * Handle SIGTERM/SIGINT gracefully
 */
function signalHandler(): void {
  // For simple plugins like this, we can just exit
  // More complex plugins would cleanup resources here
  process.exit(0);
}

/**
 * Main entry point - reads JSON from stdin, outputs JSON to stdout
 */
async function main(): Promise<void> {
  // Register signal handlers for graceful shutdown
  process.on('SIGTERM', signalHandler);
  process.on('SIGINT', signalHandler);

  try {
    // Read input from stdin
    const chunks: Buffer[] = [];

    for await (const chunk of process.stdin) {
      chunks.push(chunk as Buffer);
    }

    const inputString = Buffer.concat(chunks).toString('utf-8');
    const inputData: InputData = JSON.parse(inputString);

    // Extract parameters
    const text = inputData.text || '';
    const preserveCase = inputData.preserve_case !== undefined ? inputData.preserve_case : true;

    let result: Result;

    // Validate input
    if (!text) {
      result = {
        success: false,
        error: 'Missing required parameter: text',
        error_type: 'validation_error'
      };
    } else {
      // Execute the operation
      result = reverseString(text, preserveCase);
    }

    // Output result as JSON
    console.log(JSON.stringify(result));

  } catch (e) {
    const error = e as Error;

    // Check if it's a JSON parsing error
    const isJsonError = error.name === 'SyntaxError' && error.message.includes('JSON');

    const errorResult: ErrorResult = {
      success: false,
      error: isJsonError ? `Invalid JSON input: ${error.message}` : `Unexpected error: ${error.message}`,
      error_type: isJsonError ? 'input_error' : 'system_error'
    };

    console.log(JSON.stringify(errorResult));
    process.exit(1);
  }
}

// Run main function
main();
